sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py

```diff
@@ -14,139 +14,129 @@
 
-MetricsFn = Callable[[pl.Trainer, pl.LightningModule], Any]
-"""A metrics producing function."""
 
+Callback = TypeVar("Callback", bound=pl.Callback)
+CallbackDict = Dict[Mode, Dict[Type[pl.Callback], pl.Callback]]
 
-class MetricsDict(TypedDict):
-    """Default metrics dict."""
 
-    global_step: MetricsFn
-    learning_rate: MetricsFn
+def get_callback(callbacks: CallbackDict, mode: Mode, callback_type: Type[Callback]) -> Callback:
+    """Returns the callback with the given name and mode.
 
-
-def get_learning_rate(trainer: pl.Trainer, model: pl.LightningModule) -> Any:
-    """Returns the learning rate of the model.
+    Convenience function to make type hinting easier.
 
     Args:
-        trainer (pl.Trainer): The PyTorch Lightning trainer.
-        model (pl.LightningModule): The PyTorch Lightning model.
+        callbacks: The dictionary of callbacks.
+        mode: The mode indicating whether to stop or go.
+        callback_type: The type of the callback.
 
     Returns:
-        Any: The learning rate of the model.
+        pl.Callback: The callback with the given name and mode.
     """
-    return trainer.optimizers[0].param_groups[0]["lr"]
-
-
-def get_global_step(trainer: pl.Trainer, model: pl.LightningModule) -> Any:
-    """Returns the global step of the model.
-
-    Args:
-        trainer (pl.Trainer): The PyTorch Lightning trainer.
-        model (pl.LightningModule): The PyTorch Lightning model.
-
-    Returns:
-        Any: The global step of the model.
-    """
-    return trainer.global_step
+    return callbacks[mode][callback_type]  # type: ignore
 
 
 class StopAndGoHarness(ABC):
     """Abstract base class for a stop-and-go harness.
 
+    Users should override cls.setup_model and update cls.setUpClass to customize the downstream test cases. Metadata are
+    collected through callbacks and users can add new unit tests by comparing the metadata in stop/go stages.
+
+    By default, learning rate, global step, optimizer state, consumed samples, model weights through validation loss are
+    tested, and are accessible through cls.{stop,go}_callbacks.
+
     Stop and go tests act as follows:
-        - setup a clean model for a brief training run, select metrics to track.
-        - interrupt training via the StopAndGoException in the callback InterruptAfterMetadataCallback.
-        - setup a model to be resumed from the checkpoint, with the same metrics.
-        - Restore training and check that metadta matches the stored metrics in the callback CheckpointIntegrityCallback.
-      Useful metrics to check are things like learning rate, global step, validation loss, training loss, and anything
-        else that is important to the training process. If there is an unavailable metrics, a method for fetching the
-        metric should be provided in the bionemo.testing.callbacks module.
+        - setup a clean model for a brief training run, set StopAndGoCallback(s) to track.
```

Comment:
Rename StopAndGoCallback

Reply:
just said `callback`, I think we don't really care what they use